Chapters 1 & 2 Revisions 05-02-2025 (#5113)
dfinity/motoko: doc/md/fundamentals/1-basic-syntax/8-functions.md

# Functions

Functions in Motoko can have various attributes, the most fundamental being whether they are public or private. Public functions can be called by users or other [canisters](https://internetcomputer.org/docs/building-apps/essentials/canisters), while private functions are only accessible within the program that defines them.

The most basic Motoko [function declaration](https://internetcomputer.org/docs/motoko/fundamentals/declarations/function-declarations) is:

```motoko no-repl
func exampleFunction() : () {};
```

In objects, modules, and actors, all functions are private by default unless explicitly declared as `public`.

> ```motoko

Reviewer: this does not compile in repl, gives the error "mo:1.1-1.2: syntax error [M0002], malformed operator". either fix it so it runs in repl or add no-repl

Author: <img width="633" alt="Screenshot 2025-05-17 at 1 24 49 AM" src="https://github.com/user-attachments/assets/5c036e1f-c98b-441e-8e24-e58ef426c00b" />
Seems to work fine

> ```motoko

Reviewer: ```suggestion
```motoko no-repl
```

object Counter  {
   var value = 0;
   func reset() { value := 0 };
   public func inc() { value := 1};
   public func get() : Nat { value }; 
}
```

The object `Counter` has two public methods, the functions `Counter.inc()` and `Counter.get()`. Both `value` and `reset()` are implicitly `private`. Any attempts to access `Counter.reset()` and `Counter.value` produce type errors.

A function should specify a return type. If a return type is not declared or otherwise determined from the context, it defaults to the unit `()` return type.

> ```motoko

Reviewer: this should also have no-repl, it gives the error "mo:1.1-1.7: syntax error [M0001], unexpected token 'public', expected one of token or <phrase> sequence"

Author: This also seems to work fine

> public func exampleFunction(x : Nat) : Nat {

Reviewer 2: ```suggestion
func exampleFunction(x : Nat) : Nat {
```
(otherwise it doesn't work in the repl. Only object/module and actor members can have the keyword `public`.

Shared functions come in several forms:

- `shared` functions, which can modify an actor's state.

- `shared query` functions, which can read the actor's state without making observable changes and cannot send further messages.

- `shared composite query` functions, which behave like queries but can also call other queries.
All shared function, unlike ordinary functions, provide access to the identity of their caller, for applications like access control.

[Learn more about function types](https://internetcomputer.org/docs/motoko/fundamentals/types/functions).

:::

For example, you can rewrite the object above as an actor:

``` motoko
actor Digit {
   var value = 0;
   func reset() { value := 0 };
   public shared func inc() : async (){ 
      value += 1;
      if (value == 10) reset(); 
   };
   public shared query func get() : async Nat { value }; 
}
```

Since the public functions of an actor must be `shared`, you can omit the `shared` keyword:

``` motoko

> actor Digit {

Reviewer: Same for this actor, why not persistent?

The most basic Motoko [function declaration](https://internetcomputer.org/docs/motoko/fundamentals/declarations/function-declarations) is:

```motoko no-repl
func exampleFunction() : () {};
```

By default, all functions are declared private unless explicitly declared `public`.

```motoko no-repl

> public func exampleFunction() : () {};

Reviewer 2: ````suggestion
object Counter  {
   var value = 0;
   func reset() { value := 0 };
   public func inc() { value := 1};
   public func get() : Nat { value }; 
}
```
The object `Counter` has two public methods, the functions `Counter.inc()` and `Counter.get()`. Both `value` and `reset()` are implicitly `private` - attempts to access `Counter.reset()` and `Counter.value` are type errors.
````
But maybe it would be best to discuss `public` and `private` in `6-objects-classes.md`.

```

A function should specify a return type. If a return type is not declared, it defaults to the unit `()` return type.

```motoko no-repl
public func exampleFunction(x : Nat) : Nat {
    x;
};
```

:::info Understanding function types

Motoko functions vary by access and behaviour:

> - `private`: Not exposed outside the namespace(actor, class, module).
> - `public`: Makes the function externally callable; can be query or update.

Reviewer 2: ```suggestion
- `private`: Not exposed outside the namespace(actor, class, module).
- `public`: Makes the function externally callable; can be query or update.
```
This isn't a property of functions - just omit here.

> - `shared`: Enables the actor to identify the caller.

Reviewer: ```suggestion
- `shared`: Enables asynchronous communication between actors. The caller of a shared function can be identified and used programmatically for things such as access control. 
```

> Motoko functions vary by access and behaviour:
> 
> - `private`: Not exposed outside the namespace(actor, class, module).
> - `public`: Makes the function externally callable; can be query or update.
> - `shared`: Enables the actor to identify the caller.

Reviewer 2: ```suggestion
As well as basic functions for synchronously returning values or modifying state, Motoko functions can return futures that are executed asynchronously.

The public functions of an actor, used to send messages to the actor, are special sorts of functions called `shared` functions. Shared functions can only be declared within actors and,  unlike ordinary functions, their values can be sent to, i.e. shared, with other actors.

Shared functions come in several flavors: `shared` functions that can modify the state of an actor, `shared query` functions that can read the state of an actor without observably changing its state, but not send any further messages, and `shared composite query` functions that are similar to queries with the power to call other queries.
All shared function, unlike ordinary functions, provide access to the identity of their caller, for applications like access control.
```

- `query`: Reads data without modifying state.
- `composite query`: Reads state, can call other queries.

> 

Reviewer 2: ````suggestion
For example, we can rewrite the object above as an actor: 

``` motoko
actor Digit {
   var value = 0;
   func reset() { value := 0 };
   public shared func inc() : async (){ 
      value += 1;
      if (value == 10) reset(); 
   };
   public shared query func get() : async Nat { value }; 
}
````

The only difference is that the public functions are now declared `shared` and return their result wrapped futures (`async ()` and `async nat`)

The private `reset()` f unction can stay unchanged.

Because the public functions of an actor must be shared, you are allowed to omit the `shared` keyword:

``` motoko
actor Digit {
   var value = 0;
   func reset() { value := 0 };
   public func inc() : async () { 
      value += 1;
      if (value == 10) reset(); 
   };
   public query func get() : async Nat { value }; 
}
```
````

(If you want to add an example)

> The most basic Motoko [function declaration](https://internetcomputer.org/docs/motoko/fundamentals/declarations/function-declarations) is:

Reviewer 2: Actually, public/private isn't an attribute of functions (unlike shared/query etc), but only of the declarations within object/modules and actors. I'd actually delete this sentence.

```motoko no-repl
func exampleFunction() : () {};
```

> Within objects, modules and actors all functions are declared private unless explicitly declared `public`.

Reviewer 2: Objects, modules and actors are actually different things with different types, so the original sentence is IMO better.

> Within objects, modules and actors all functions are declared private unless explicitly declared `public`.

Reviewer: Okay, the original sentence is not clear or grammatically correct, IMO. How about this instead, then?

```
In objects, modules, and actors, all functions are private by default unless explicitly declared as public.
```

```motoko no-repl
object Counter  {
   var value = 0;
   func reset() { value := 0 };
   public func inc() { value := 1};
   public func get() : Nat { value }; 
}
```

The object `Counter` has two public methods, the functions `Counter.inc()` and `Counter.get()`. Both `value` and `reset()` are implicitly `private` - attempts to access `Counter.reset()` and `Counter.value` are type errors.

A function should specify a return type. If a return type is not declared or otherwise determined from the context, it defaults to the unit `()` return type.

```motoko no-repl
public func exampleFunction(x : Nat) : Nat {
    x;
};
```

:::info Understanding function types

Motoko functions vary by access and behaviour:

> The public functions of an actor, used to send messages to the actor, are special sorts of functions called `shared` functions. Shared functions can only be declared within actors and,  unlike ordinary functions their values can be sent to, that is _shared with_, other actors.

Reviewer: This did not get applied.

> Shared functions come in several flavors: `shared` functions that can modify the state of an actor, `shared query` functions that can read the state of an actor without observably changing its state, but not send any further messages, and `shared composite query` functions that are similar to queries with the power to call other queries.

Reviewer: This also did not get applied

All shared function, unlike ordinary functions, provide access to the identity of their caller, for applications like access control.

[Learn more about functions](https://internetcomputer.org/docs/motoko/fundamentals/types/functions).

:::

For example, we can rewrite the object above as an actor:

``` motoko
actor Digit {
   var value = 0;
   func reset() { value := 0 };
   public shared func inc() : async (){ 
      value += 1;
      if (value == 10) reset(); 
   };
   public shared query func get() : async Nat { value }; 
}
```

> The only difference is that the public functions are now declared `shared` and return their result wrapped futures (`async ()` and `async nat`)

Reviewer 2: ```suggestion
The only difference is that the public functions are now declared `shared` and return their results wrapped in futures (`async ()` and `async nat`)
```

> The private `reset()` f unction can stay unchanged.

Reviewer 2: ```suggestion
The private `reset()` function can stay unchanged.
```
